---
title: "Dynamics"
date: "`r format(Sys.time(), '%Y %m %d')`"
output:
  html_document:
    code_folding: hide
    theme: united
    toc: yes
    toc_float: yes
---

# Setup


Load packages & original functions

- `summarise_simul` : Load & output simulation result
- `make_config` : Make simulation config file

```{r}

rm(list=ls())  # reset workspace

pacman::p_load(
 openxlsx    # open excel file
 , RcppRoll  # window functions
 , tidyverse
 , reshape2
 , foreach   # Pararel
 , magrittr  # Extend pipe
 , knitr     # rmarkdown
 , kableExtra # nice tables
 , pander     # nice tables
 , patchwork
 )

opts_chunk$set(prompt=TRUE, message=FALSE, comment="", warning = FALSE) 
options(knitr.kable.NA = '', scipen=100)  # do not show numbers using exponential


#' Title: make_config
#' Make simulation configuration file
#' @param gset: Genome setting. It has to be either [dis_unlink, dis_link, quant_learn, quant]
#' @param sexsel: Sexual selection style
#' @param gen: Generations
#' @param mig: Migration in each >1 generation
#' @param maten: Sampled mate
make_config <- function(
  gset = "", sexsel = "", natsel = "", gen = 100, mig = 0, maten = 50
  ) {

  paste(
    "RandomSeed\t=\t.1356",         
    paste0("MarkerFile\t=\t/lustre/k.atsumi/Config/markers_", gset, ".txt"),
    # "MarkerFile\t=\tmarkers.txt",    #Marker File Containing marker info
    paste0("MarkerProbFile\t=\t/lustre/k.atsumi/Config/recomb_", gset, ".txt"),
    # "MarkerProbFile\t=\trecomb.txt", #Recombination probabilities
    "UseUniformRec\t=\tyes",		# ignore predefined breakpoints above, use uniform.
    "IgnoreMarkerFreq\t=\tyes", #setting this option to yes will treat all markers as ancestry informative, regardless of their frequency specified in the markers file
    # "DumpNatSelProb\t=\tOff", 
    paste0("GeneFile\t=\t/lustre/k.atsumi/Config/genes_", gset, ".txt"),
    # "GeneFile\t=\t/lustre/k.atsumi/Config/genes.txt",
    paste0("PhenotypeFile\t=\t/lustre/k.atsumi/Config/phenotypes_", gset, ".txt"),
    paste0("OutputFolder\t=\t/lustre/k.atsumi/Result/", gset, "_", sexsel, "_natsel", natsel, "/0"), #Output folder
    paste0("NaturalSelection\t=\t/lustre/k.atsumi/Config/naturalsel_", natsel, ".txt"),                     #Natural selection file
    paste0("SexualSelection\t=\t/lustre/k.atsumi/Config/sexualsel_", sexsel, ".txt"), #Sexual selection file
    #population names:
    "pop2_name\t=\tp2",
    "pop2_ancestry_label\t=\tp2",
    "hybrid_name\t=\thyb",
    "pop1_name\t=\tp1",
    "pop1_ancestry_label\t=\tp1",
    #Pop sizes
    "pop1_init_size\t=\t500",
    "pop1_size_limit\t=\t500",
    "pop2_init_size\t=\t500",
    "pop2_size_limit\t=\t500",
    "hybrid_size_limit\t=\t2000",
    #Sex ratio
    "pop1_male_ratio\t=\t0.5",
    "pop2_male_ratio\t=\t0.5",
    #Migration
    "migration_only_first_gen\t=\tyes",
    #migration first generation:
    "gen0_pop2_to_hybrid\t=\t500",
    "gen0_pop1_to_hybrid\t=\t500",
    #migration each gen:
    paste0("pop2_to_pop1\t=\t", mig),
    paste0("pop2_to_pop1\t=\t", mig),
    paste0("pop2_to_hybrid\t=\t", mig),
    paste0("hybrid_to_pop2\t=\t", mig),
    paste0("hybrid_to_pop1\t=\t", mig),
    paste0("pop1_to_pop2\t=\t", mig),
    paste0("pop1_to_hybrid\t=\t", mig),
    paste0("generations\t=\t", gen),
    "avg_female_gamete\t=\t10",
    "std_female_gamete\t=\t3.16",
    "kids_per_female_func\t=\tPoisson",
    "samplegens\t=\t/lustre/k.atsumi/Config/samplegens.txt", 
    "MarkerOutput\t=\tOff", #On or Off
    paste0("SampleMate\t=\t", maten),
    "NumThreads\t=\t1",
    "DisableOutputForFirst3Gens\t=\tyes",
    sep = "\n"
    ) %>% 
    return()
  
}


#' Title: make_bash
#' Make simulation bash file
#' @param gset: Genome setting. It has to be either [dis_unlink, dis_link, quant_learn, quant]
#' @param sexsel: Sexual selection style
#' @param gen: Generations
#' @param mig: Migration in each >1 generation
#' @param maten: Sampled mate
make_bash <- function(gset = "", sexsel = "") {
  
  paste(
    "#!/bin/bash",
    "#PBS -N hybrid_simulation",
    "#PBS -q blade",
    "#PBS -l nodes=1:ppn=3,mem=50gb,walltime=96:00:00",
    "module load gcc",
    "REPS=101",
    "COUNTERA=0",
    "COUNTERB=0",
    "X=.1356",
    "Y=.1356",
    "while [ $COUNTERA -lt $REPS ]; do",
    " echo \"$COUNTERA\"",
    " Y=`echo \"$Y + 0.0001\" | bc`",
    " echo $Y",
    paste0(" sed -i -e \"s/$X/$Y/g\" /lustre/k.atsumi/Config/", gset, "_", sexsel, "_natsel0.1.cfg"),
    paste0(" sed -i -e \"s/$X/$Y/g\" /lustre/k.atsumi/Config/", gset, "_", sexsel, "_natsel0.4.cfg"),
    paste0(" sed -i -e \"s/$X/$Y/g\" /lustre/k.atsumi/Config/", gset, "_", sexsel, "_natsel0.7.cfg"),
    paste0(" sed -i -e \"s/$X/$Y/g\" /lustre/k.atsumi/Config/", gset, "_", sexsel, "_natsel1.cfg"),
    " let COUNTERB=COUNTERB+1",
    paste0(" sed -i -e \"s/rep${COUNTERA}dir/rep${COUNTERB}dir/g\" /lustre/k.atsumi/Config/", gset, "_", sexsel, "_natsel0.1.cfg"),
    paste0(" sed -i -e \"s/rep${COUNTERA}dir/rep${COUNTERB}dir/g\" /lustre/k.atsumi/Config/", gset, "_", sexsel, "_natsel0.4.cfg"),
    paste0(" sed -i -e \"s/rep${COUNTERA}dir/rep${COUNTERB}dir/g\" /lustre/k.atsumi/Config/", gset, "_", sexsel, "_natsel0.7.cfg"),
    paste0(" sed -i -e \"s/rep${COUNTERA}dir/rep${COUNTERB}dir/g\" /lustre/k.atsumi/Config/", gset, "_", sexsel, "_natsel1.cfg"),
    " echo $X",
    paste0(" /lustre/k.atsumi/admixem/bin/admixemp /lustre/k.atsumi/Config/", gset, "_", sexsel, "_natsel0.1.cfg"),
    paste0(" /lustre/k.atsumi/admixem/bin/admixemp /lustre/k.atsumi/Config/", gset, "_", sexsel, "_natsel0.4.cfg"),
    paste0(" /lustre/k.atsumi/admixem/bin/admixemp /lustre/k.atsumi/Config/", gset, "_", sexsel, "_natsel0.7.cfg"),
    paste0(" /lustre/k.atsumi/admixem/bin/admixemp /lustre/k.atsumi/Config/", gset, "_", sexsel, "_natsel1.cfg"),
    " X=`echo \"$X + 0.0001\" | bc`",
    " let COUNTERA=COUNTERA+1",
    "done",
    "wait",
    sep = "\n"
    ) %>% 
    return()
  
  }
  

# Make inherit style table
inherit_table <- tibble(
  "rowname" = c("TT", "Tt", "tt"),
  "PP" = NA, "Pp" = NA, "pp" = NA
  ) %>% 
  column_to_rownames("rowname")


#' Title: load_simul
#' Load simulation result
#' @locus l: Number of locus
#' @param sexsel: Sexual selection style
#' @param gen: Generations to simulate
load_simul <- function(l = "", sexsel = "", gen = 100, natsel = "") {
  
  gens <- c(5, seq(10, gen, by = 10))

  # Load simulation data. Repeat for all generation files
  foreach(g = gens, .combine = 'rbind') %do% {
    # Read tab delimited files for phenotype
    read_tsv(paste0("./Result/", l, "_", sexsel, "_natsel", natsel, "/0/Gen", g, "_phenotypes.txt")) %>%
      # Use only hybrid population
      filter(pop == 3) %>%
      dplyr::select(contains("Signal")) %>% #id,Sex,contains("Pref"),-pop
      # Add generation
      mutate(Generation = g)
    } %>% 
    mutate(
      Locus = locus %>% as.factor,
      Naturalselection = str_c("Selection =", ns)
      )
  
}


#' Title: summarise_simul
#' Make simulation configuration file
#' @param d: Simulation data loaded by `load_simul` function
summarise_simul <- function(d = "") {
  
  d %<>%  as.tibble(d) %>%  
    mutate_at(vars(contains("Signal")), as.numeric) %>% 
    mutate_all(as.factor)
  
  left_join(
    # Count individuals grouped by generation and courter phenotype
    d %>% 
      group_by(Generation, Signal0, Naturalselection, Locus) %>% 
      summarise(nPheno = length(Generation)), 
    # Count individuals grouped by generation
    d %>%
      group_by(Generation, Naturalselection, Locus) %>% 
      summarise(nGen = length(Generation))
    ) %>% 
    mutate(Freq = nPheno/nGen*100) %>% 
    mutate_at("Signal0", as.factor) %>% 
    ungroup() %>% 
    return()
}

# Plot settings
mytheme <- theme_bw() + # simple theme
  theme(panel.grid.major.x = element_blank(),
        axis.title = element_text(size=12),
        axis.text = element_text(size=11),
        axis.text.x = element_text(angle = 30, hjust = 1),
        strip.text = element_text(size=12, lineheight=5.0),
        strip.background = element_rect(fill = "lightgrey")
        ) 

cbPalette <- c("#999999", "#E69F00", "#56B4E9", "#009E73", "#0072B2", "#D55E00", "#CC79A7")

```


# Model settings

- With sex chromosome 1
  - Allele (S) with trait value 1 will be male

- 100% mortal epistasis between Chrom 2 & 3

- Migration happens in only first generation

- Max 2000 indv for hybrid pop

- Mate sampling: 50 indv

Here I set original function `make_genome` that creates genes file

```{r}

chrom_size <- 1*10^5
paste("Chromosone size (bp):", chrom_size)

g_chrom <- 2
paste("Genes per chromosome:", g_chrom)


#' Title: make_genome
#' Make genes file
#' @param chrom_n: Number of total chromosomes
#' @param g_chrom: Genes/chromosome
#' @param g_total: Total number of genes
#' @param chrom_size: Chromosome size in basepair
make_genome <- function(cn = chrom_n, gt = g_total) {

  set.seed(10) # To make randomization repeatable
  
  gf <- data.frame(
    Chromosome = c(
      1,                      # Sex chromosome
      rep(2:cn, g_chrom) # Autosomal chromosome
      ) %>% 
      sort(),
    # Create random chromosome position to each autosomal gene loci
    position = runif(n = gt, min = 1, max = chrom_size) %>% 
        floor(), # Integers
    # First gene: Sex determination. S vs s. Hemizygous. Freq = 0.5
    Allele1     = c("S", rep("I", 2*g_chrom), rep("A", gt - (2*g_chrom+1))),
    Allele1val  = c(1, rep(0.5, g_total - 1)),
    Allele2     = c("s", rep("i", 2*g_chrom), rep("a", gt - (2*g_chrom+1))),
    Allele2val  = c(0, rep(0, gt - 1)),
    Mode        = c("Hemizygous", rep("Additive", gt - 1)),
    Freqal1pop1 = c(0.5, rep(1.0, gt - 1)),
    Freqal1pop2 = c(0.5, rep(0, gt - 1))
    ) %>% 
    mutate(name = str_c("chr", Chromosome, "_", position, sep = "")) %>% 
    dplyr::select(name, Chromosome:Freqal1pop2)
  
  # Epistasis
  gf[2, "Allele1val"] <- 5
  gf[2, "Allele2val"] <- 0
  gf[2+g_chrom, "Allele1val"] <- 0
  gf[2+g_chrom, "Allele2val"] <- 10
  
  return(gf)

}

```


## Simple models

```{r}

g_trait_n <- 1    # N of trait loci
paste(g_trait_n, "locus determine courter trait")

g_pref_n  <- 1     # N of preference loci
paste(g_pref_n, "locus determine mate preference peak, with constant choosiness")
# trait_n   <- 1      # N of courter trait

chrom_n <- 1 + 2 + 10 + g_trait_n + g_pref_n #Sex, BDMI, Hybrid index, Trait, Preference (Maximum number)
paste("Chromosone number (with sex chrom):", chrom_n)

g_total <- 1 + (chrom_n - 1)*g_chrom
paste("Total number of genes:", g_total)

# Create gene file
gf <- make_genome()
write_tsv(gf, "./Config/genes_dis_link.txt")
file.copy(
  "./Config/genes_dis_link.txt", "./Config/genes_dis_unlink.txt", overwrite = T
  )

```


### Create phenotype files

Create `phenotypes_{genome setting}.txt`. I made 3 phenotype files with genome setting
- *dis_unlink* : Trait determined by 1 locus, which is unlinked with preference genes
- *dis_link* : Trait determined by 1 locus, which is same gene determining peak

```{r}

Hybridindex <- paste0(
  gf[2+g_chrom*2, "name"], "+", gf[3+g_chrom*2, "name"], "+", 
  gf[2+g_chrom*3, "name"], "+", gf[3+g_chrom*3, "name"], "+", 
  gf[2+g_chrom*4, "name"], "+", gf[3+g_chrom*4, "name"], "+", 
  gf[2+g_chrom*5, "name"], "+", gf[3+g_chrom*6, "name"], "+", 
  gf[2+g_chrom*6, "name"], "+", gf[3+g_chrom*6, "name"], "+", 
  gf[2+g_chrom*7, "name"], "+", gf[3+g_chrom*7, "name"], "+", 
  gf[2+g_chrom*8, "name"], "+", gf[3+g_chrom*8, "name"], "+", 
  gf[2+g_chrom*9, "name"], "+", gf[3+g_chrom*9, "name"], "+",
  gf[2+g_chrom*10, "name"], "+", gf[3+g_chrom*10, "name"], "+",
  gf[2+g_chrom*11, "name"], "+", gf[3+g_chrom*11, "name"]
  )

Pheno <- data.frame(
  Phenotypes = c("Sex", "Fitness0", "Pref0", "Pref1", "Signal0", "Hybridindex"), 
  Formula    = c(
    gf[1, "name"],   # Sex
    paste0(gf[2, "name"], "+", gf[2+g_chrom, "name"]), # Fitness0 = B
    gf[2+g_chrom*12, "name"], # Preference (peak)
    gf[2+g_chrom*13, "name"], # Preference (choosiness)
    "Signal",                # Signal
    Hybridindex              # Hybrid Index
    )
  )

# dis_unlink
Pheno[5,2] <- gf[2+g_chrom*14, "name"]
write_tsv(Pheno, "./Config/phenotypes_dis_unlink.txt")

# dis_link
Pheno[5,2] <- gf[2+g_chrom*2, "name"]
write_tsv(Pheno, "./Config/phenotypes_dis_link.txt")


for(genomeset in c("dis_unlink", "dis_link")){
  for(ss in c("speriorhyb", "disruptive", "flip", "heterorandom", "pdominance", "recessive", "selfref")){
    make_bash(gset = genomeset, sexsel = ss) %>% 
      write(paste0("./Bash/", genomeset, "_", ss, ".sbatch.sh"))
    
    for(ns in c(0.1, 0.4, 0.7, 1)){
      make_config(gset = genomeset, sexsel = ss, natsel = ns) %>% 
        write(paste0("./Config/", genomeset, "_", ss, "_natsel", ns, ".cfg"))
      
      # dir.create(paste0("/Users/katsu/Dropbox/5_Simulation/Result/", genomeset, "_", ss, "_natsel", ns)) # #出力先フォルダ作成
    }}}

# By running this file on Admix'em, marker file (`markers.txt`) and recombination file (`recomb.txt`) will be made

m_chrom <- 2 #Marker per chromosome

paste(
  "1",        #Generate markers
  "0.99",     #Random seed
  chrom_n,    #N of Chromosome 
  sep = "\n"  #改行
  ) %>% 
  cat(
    ., "\n", file = "./Config/config_marker_dis_link.txt", 
    append = FALSE  #Renew file
    )

for(c in 1:chrom_n){　#Repeat chromosome size
  cat(
    chrom_size, "\n", file = "./Config/config_marker_dis_link.txt",
    append = TRUE   #Add to existence file
    )
  }

paste(
  chrom_n*m_chrom,   #N of markers
  "1.0",               #Pop 1 allele freq
  0,                 #Pop 1 allele freq SD
  "1.0",               #Pop 2 allele freq
  0,                 #Pop 2 allele freq SD
  "markers_dis_link.txt",     #marker file name
  "y",                 #Reconbination file
  4,                 #Reconbination N
  5,                 #Male female difference
  "recomb_dis_link.txt",      #recombination file
  sep = "\n"
  ) %>% 
  cat(., "\n", file = "./Config/config_marker_dis_link.txt", append = TRUE)

file.copy("./Config/recomb_dis_link.txt","./Config/recomb_dis_unlink.txt", overwrite = T)
file.copy("./Config/markers_dis_link.txt","./Config/markers_dis_unlink.txt", overwrite = T)

```


## Quantitative trait & Learned preference models

Preference is determined not by gene, but by phenotypic distribution of the hybrid population

```{r}

g_trait_n <- 16    # N of trait loci
paste(g_trait_n, "locus determine courter trait")

g_pref_n  <- 0     # N of preference loci
paste(g_pref_n, "locus determine mate preference peak, with constant choosiness")
# trait_n   <- 1      # N of courter trait

chrom_n <- 1 + 2 + 10 + g_trait_n + g_pref_n #Sex, BDMI, Hybrid index, Trait, Preference (Maximum number)
paste("Chromosone number (with sex chrom):", chrom_n)

g_total <- 1 + (chrom_n - 1)*g_chrom
paste("Total number of genes:", g_total)

# Create gene file
gf <- make_genome()
write_tsv(gf, "./Config/genes_learn.txt")

# Phenotype file
Pheno[5,2] <- paste0(
  gf[2+g_chrom*12, "name"], "+", gf[2+g_chrom*13, "name"], "+",   
  gf[2+g_chrom*14, "name"], "+", gf[2+g_chrom*15, "name"], "+", 
  gf[2+g_chrom*16, "name"], "+", gf[2+g_chrom*17, "name"], "+", 
  gf[2+g_chrom*18, "name"], "+", gf[2+g_chrom*19, "name"], "+", 
  gf[2+g_chrom*20, "name"], "+", gf[2+g_chrom*21, "name"], "+", 
  gf[2+g_chrom*22, "name"], "+", gf[2+g_chrom*23, "name"], "+", 
  gf[2+g_chrom*24, "name"], "+", gf[2+g_chrom*25, "name"], "+", 
  gf[2+g_chrom*26, "name"], "+", gf[2+g_chrom*27, "name"]
  )
write_tsv(Pheno, "./Config/phenotypes_learn.txt")  

# Config file
for(ss in c("obl_posit", "obl_negat")){
  make_bash(gset = "learn", sexsel = ss) %>% 
      write(paste0("./Bash/", "learn_", ss, ".sbatch.sh"))

  for(ns in c(0.1, 0.4, 0.7, 1)){
    make_config(gset = "learn", sexsel = ss, natsel = ns) %>% 
      write(paste0("./Config/learn_", ss, "_natsel", ns, ".cfg"))
    # #出力先フォルダ作成
    # dir.create(paste0("/Users/katsu/Dropbox/5_Simulation/Result/learn_", ss, "_natsel", ns))
  }}


# By running this file on Admix'em, marker file (`markers.txt`) and recombination file (`recomb.txt`) will be made
m_chrom <- 2 #Marker per chromosome

paste(
  "1",        #Generate markers
  "0.99",     #Random seed
  chrom_n,    #N of Chromosome 
  sep = "\n"  #改行
  ) %>% 
  cat(
    ., "\n", file = "./Config/config_marker_learn.txt", 
    append = FALSE  #Renew file
    )

for(c in 1:chrom_n){　#Repeat chromosome size
  cat(
    chrom_size, "\n", file = "./Config/config_marker_learn.txt",
    append = TRUE   #Add to existence file
    )
  }

paste(
  chrom_n*m_chrom,   #N of markers
  "1.0",               #Pop 1 allele freq
  0,                 #Pop 1 allele freq SD
  "1.0",               #Pop 2 allele freq
  0,                 #Pop 2 allele freq SD
  "markers_learn.txt",     #marker file name
  "y",                 #Reconbination file
  4,                 #Reconbination N
  5,                 #Male female difference
  "recomb_learn.txt",      #recombination file
  sep = "\n"
  ) %>% 
  cat(., "\n", file = "./Config/config_marker_learn.txt", append = TRUE)

```

## Quantitative trait and preference

```{r}

g_trait_n <- 16    # N of trait loci
paste(g_trait_n, "loci determine courter trait")

g_pref_n  <- 16     # N of preference loci
paste(g_pref_n, "loci determine mate preference peak, with constant choosiness")
# trait_n   <- 1      # N of courter trait

chrom_n <- 1 + 2 + 10 + g_trait_n + g_pref_n #Sex, BDMI, Hybrid index, Trait, Preference (Maximum number)
paste("Chromosone number (with sex chrom):", chrom_n)

g_total <- 1 + (chrom_n - 1)*g_chrom
paste("Total number of genes:", g_total)

# Genes file
gf <- make_genome()
write_tsv(gf, "./Config/genes_quant.txt")

# Phenotype file
Pheno[3,2] <- paste0(
  gf[2+g_chrom*28, "name"], "+", gf[2+g_chrom*29, "name"], "+", 
  gf[2+g_chrom*30, "name"], "+", gf[2+g_chrom*31, "name"], "+", 
  gf[2+g_chrom*32, "name"], "+", gf[2+g_chrom*33, "name"], "+", 
  gf[2+g_chrom*34, "name"], "+", gf[2+g_chrom*35, "name"], "+", 
  gf[2+g_chrom*36, "name"], "+", gf[2+g_chrom*37, "name"], "+", 
  gf[2+g_chrom*38, "name"], "+", gf[2+g_chrom*39, "name"], "+", 
  gf[2+g_chrom*40, "name"], "+", gf[2+g_chrom*41, "name"], "+", 
  gf[2+g_chrom*42, "name"], "+", gf[2+g_chrom*43, "name"]
  )
write_tsv(Pheno, "./Config/phenotypes_quant.txt")  

# Config file
for(ss in c("constchoosy", "hybridrandom")){
  make_bash(gset = "quant", sexsel = ss) %>% 
      write(paste0("./Bash/", "quant_", ss, ".sbatch.sh"))
  
  for(ns in c(0.1, 0.4, 0.7, 1)){
    make_config(gset = "quant", sexsel = ss, natsel = ns) %>%
      write(paste0("./Config/quant_", ss, "_natsel", ns, ".cfg"))
    # #出力先フォルダ作成
    # dir.create(paste0("/Users/katsu/Dropbox/5_Simulation/Result/quant_", ss, "_natsel", ns))
  }}

# By running this file on Admix'em, marker file (`markers.txt`) and recombination file (`recomb.txt`) will be made

m_chrom <- 2 #Marker per chromosome

paste(
  "1",        #Generate markers
  "0.99",     #Random seed
  chrom_n,    #N of Chromosome 
  sep = "\n"  #改行
  ) %>% 
  cat(
    ., "\n", file = "./Config/config_marker_quant.txt", 
    append = FALSE  #Renew file
    )

for(c in 1:chrom_n){　#Repeat chromosome size
  cat(
    chrom_size, "\n", file = "./Config/config_marker_quant.txt",
    append = TRUE   #Add to existence file
    )
  }

paste(
  chrom_n*m_chrom,   #N of markers
  "1.0",               #Pop 1 allele freq
  0,                 #Pop 1 allele freq SD
  "1.0",               #Pop 2 allele freq
  0,                 #Pop 2 allele freq SD
  "markers_quant.txt",     #marker file name
  "y",                 #Reconbination file
  4,                 #Reconbination N
  5,                 #Male female difference
  "recomb_quant.txt",      #recombination file
  sep = "\n"
  ) %>% 
  cat(., "\n", file = "./Config/config_marker_quant.txt", append = TRUE)

```

- *quant_unlink* : Trait determined by independent 16 loci which are unlinked with preference genes

<div style="margin-bottom:70px;">
</div>


### Bash file

```

#!/bin/bash
#PBS -N hybrid_simulation
#PBS -q blade
#PBS -l nodes=1:ppn=1,mem=50gb,walltime=96:00:00

# we are going to do from repetition 1 to 10
# X is going to be the first random seed value, its decimal because random seeds are between 0 and 1. If there is an existing one, let X be the newest one.
# Y is goin to be the new randmon seed value; let it have the same value as X here

# GNUコンパイラ
module load gcc

REPS=101
COUNTERA=0
COUNTERB=0
X=.1356
Y=.1356

# -lt : less than
#$COUNTER<$REPSである限り繰り返す
while [ $COUNTERA -lt $REPS ]; do

   echo "$COUNTERA"
   #新Random seed値をYに代入
   #bc : 小数を扱える計算処理コマンド。echo 計算式 | bc
   Y=`echo "$Y + 0.0001" | bc`
   echo $Y

   #ConfigファイルのRandom seedを書き換え
   sed -i -e "s/$X/$Y/g" <config file path & name>

   let COUNTERB=COUNTERB+1
   #Configファイルの出力先フォルダを書き換え
   #rep{Number}dir → rep{Number+1}dir
   sed -i -e "s/rep${COUNTERA}dir/rep${COUNTERB}dir/g" /lustre/k.atsumi/Config/dis_link_disruptive_natsel0.3.cfg
   echo $X

   #Run simulation
   <admixemp file path> <config file path & name> 

   #X becomes the next random seed number
   X=`echo "$X + 0.0001" | bc`

   #COUNTERを1つ進める
   #let : 整数のみの四則演算や論理演算を含む算術式を評価
   let COUNTERA=COUNTERA+1

done
wait

```



<div style="margin-bottom:70px;">
</div>

## Result

- 2 loci models : Preference and traits are determined by each 1 gene on different chromosome

- 1 locus models : Preference and traits are determined by 1 gene

### Disruptive sexual selection

```{r}

# Visualize inheritance pattern
inherit_table$PP <- c(0.745,0.01,0.245)
inherit_table$Pp <- c(0.495,0.01,0.495)
inherit_table$pp <- c(0.245,0.01,0.745)
inherit_table %>% 
  kable("html", caption = "**Trait \\ Preference**") %>% 
  kable_styling("striped", position = "left", full_width = F)

cap = "Disatractive hybrids"

dat <- foreach(locus = 1:2, .combine = 'rbind') %do% {
  foreach(ns = c(0.3, 1), .combine = 'rbind') %do% {
    load_simul(l = locus, sexsel = "disruptive", natsel = ns)
    } } %>% 
  summarise_simul()
levels(dat$Locus) <- c("Linked", "Unlinked")

dat %>% ggplot(aes(x = Generation, y = Freq, colour = Signal0, group = Signal0)) +
  geom_line() +
  labs(y = "Frequency (%)", colour = "Courter phenotype", title = cap) +
  facet_grid(Naturalselection ~ Locus, scales = "free")

```

<div style="margin-bottom:40px;">
</div>


### Self-reference assortative matings 

```{r}
# Visualize inheritance pattern
inherit_table$PP <- c(0.99,0.005,0.005)
inherit_table$Pp <- c(0.005,0.99,0.005)
inherit_table$pp <- c(0.005,0.005,0.99)
# Output inheritance style table
inherit_table %>% 
  kable("html", caption = "**Trait \\ Preference**") %>% 
  # here you can add the vertical line
  column_spec (2:4, border_left = T, border_right = T) %>%
  kable_styling(position = "left", full_width = F)

cap = "Self-reference assortative mating"

dat <- foreach(locus = 1:2, .combine = 'rbind') %do% {
  foreach(ns = c(0.3, 1), .combine = 'rbind') %do% {
    load_simul(l = locus, sexsel = "selfref", natsel = ns) 
    } } %>% 
  summarise_simul()
levels(dat$Locus) <- c("Linked", "Unlinked")

dat %>% ggplot(aes(x = Generation, y = Freq, colour = Signal0, group = Signal0)) +
  geom_line() +
  labs(y = "Frequency (%)", colour = "Courter phenotype", title = cap) +
  facet_grid(Naturalselection ~ Locus, scales = "free")

```

<div style="margin-bottom:40px;">
</div>

### Recessive preference

```{r}

# Visualize inheritance pattern
inherit_table$PP <- c(0.99,0.005,0.005)
inherit_table$Pp <- c(0.005,0.005,0.99)
inherit_table$pp <- c(0.005,0.005,0.99)
inherit_table %>% 
  kable("html", caption = "**Trait \\ Preference**") %>% 
  # here you can add the vertical line
  column_spec (2:4, border_left = T, border_right = T) %>%
  kable_styling(position = "left", full_width = F)
cap = "Recessive preference"

dat <- foreach(locus = 1:2, .combine = 'rbind') %do% {
  foreach(ns = c(0.3, 1), .combine = 'rbind') %do% {
    load_simul(l = locus, sexsel = "recessive", natsel = ns)
    } } %>% 
  summarise_simul()
levels(dat$Locus) <- c("Linked", "Unlinked")

dat %>% ggplot(aes(x = Generation, y = Freq, colour = Signal0, group = Signal0)) +
  geom_line() +
  labs(y = "Frequency (%)", colour = "Courter phenotype", title = cap) +
  facet_grid(Naturalselection ~ Locus, scales = "free")

```

<div style="margin-bottom:40px;">
</div>

### Pseudo-dominance

```{r}

# Visualize inheritance pattern
inherit_table$PP <- c(0.99,0.005,0.005)
inherit_table$Pp <- c(0.99,0.005,0.005)
inherit_table$pp <- c(1/3,1/3,1/3)
inherit_table %>% 
  kable("html", caption = "**Trait \\ Preference**") %>% 
  kable_styling("striped", position = "left", full_width = F)

cap = "Pseudo-dominance"

dat <- foreach(locus = 1:2, .combine = 'rbind') %do% {
  foreach(ns = c(0.3, 1), .combine = 'rbind') %do% {
    load_simul(l = locus, sexsel = "pdominance", natsel = ns)
    } } %>% 
  summarise_simul()
levels(dat$Locus) <- c("Linked", "Unlinked")

dat %>% ggplot(aes(x = Generation, y = Freq, colour = Signal0, group = Signal0)) +
  geom_line() +
  labs(y = "Frequency (%)", colour = "Courter phenotype", title = cap) +
  facet_grid(Naturalselection ~ Locus, scales = "free")

```

Heterozygotes and recessive are favored (why??)

<div style="margin-bottom:40px;">
</div>

### Heterozygotes mate randomly

```{r}

# Visualize inheritance pattern
inherit_table$PP <- c(0.99,0.005,0.005)
inherit_table$Pp <- c(1/3,1/3,1/3)
inherit_table$pp <- c(0.005,0.005,0.99)
inherit_table %>% 
  kable("html", caption = "**Trait \\ Preference**") %>% 
  kable_styling("striped", position = "left", full_width = F)

cap = "Heterozygotes mate randomly"

dat <- foreach(locus = 1:2, .combine = 'rbind') %do% {
  foreach(ns = c(0.3, 1), .combine = 'rbind') %do% {
    load_simul(l = locus, sexsel = "heterorandom", natsel = ns)
    } } %>% 
  summarise_simul()
levels(dat$Locus) <- c("Linked", "Unlinked")

dat %>% ggplot(aes(x = Generation, y = Freq, colour = Signal0, group = Signal0)) +
  geom_line() +
  labs(y = "Frequency (%)", colour = "Courter phenotype", title = cap) +
  facet_grid(Naturalselection ~ Locus, scales = "free")

```

Recessive is favored, and heterozygote is not favored

<div style="margin-bottom:40px;">
</div>

### Flip preference

```{r}

# Visualize inheritance pattern
inherit_table$PP <- c(0.99,0.005,0.005)
inherit_table$Pp <- c(0.49,0.01,0.49)
inherit_table$pp <- c(0.005,0.005,0.99)
inherit_table %>% 
  kable("html", caption = "**Trait \\ Preference**") %>% 
  kable_styling("striped", position = "left", full_width = F)

cap = "Flip preference"

dat <- foreach(locus = 1:2, .combine = 'rbind') %do% {
  foreach(ns = c(0.3, 1), .combine = 'rbind') %do% {
    load_simul(l = locus, sexsel = "flip", natsel = ns)
    } } %>% 
  summarise_simul()
levels(dat$Locus) <- c("Linked", "Unlinked")

dat %>% ggplot(aes(x = Generation, y = Freq, colour = Signal0, group = Signal0)) +
  geom_line() +
  labs(y = "Frequency (%)", colour = "Courter phenotype", title = cap) +
  facet_grid(Naturalselection ~ Locus, scales = "free")

```

<div style="margin-bottom:40px;">
</div>

### Atractive hybrids

```{r}

# Visualize inheritance pattern
inherit_table$PP <- c(0.245,0.745,0.01)
inherit_table$Pp <- c(0.005,0.99,0.005)
inherit_table$pp <- c(0.01,0.745,0.245)
inherit_table %>% 
  kable("html", caption = "**Trait \\ Preference**") %>% 
  kable_styling("striped", position = "left", full_width = F)

cap = "Atractive hybrids"

dat <- foreach(locus = 1:2, .combine = 'rbind') %do% {
  foreach(ns = c(0.3, 1), .combine = 'rbind') %do% {
    load_simul(l = locus, sexsel = "speriorhyb", natsel = ns)
    } } %>% 
  summarise_simul()
levels(dat$Locus) <- c("Linked", "Unlinked")

dat %>% ggplot(aes(x = Generation, y = Freq, colour = Signal0, group = Signal0)) +
  geom_line() +
  labs(y = "Frequency (%)", colour = "Courter phenotype", title = cap) +
  facet_grid(Naturalselection ~ Locus, scales = "free")

```

<div style="margin-bottom:40px;">
</div>

### Oblique positively learned preference

```{r}

cap = "Oblique positive"

dat <- foreach(ns = c(0.3, 1), .combine = 'rbind') %do% {
  load_simul(l = 8, sexsel = "obl_posit", natsel = ns)
  } %>% 
  summarise_simul()

dat %>% ggplot(aes(x = Generation, y = Signal0, size = Freq)) +
  geom_point() +
  labs(y = "Frequency (%)", title = cap) +
  facet_grid(Naturalselection ~ Locus, scales = "free")

```

<div style="margin-bottom:40px;">
</div>

### Oblique negatively learned preference

```{r}

cap = "Oblique negative"

dat <- foreach(ns = c(0.3, 1), .combine = 'rbind') %do% {
  load_simul(l = 8, sexsel = "obl_negat", natsel = ns)
    } %>% 
  summarise_simul()

dat %>% ggplot(aes(x = Generation, y = Signal0, size = Freq)) +
  geom_point() +
  labs(y = "Frequency (%)", title = cap) +
  facet_grid(Naturalselection ~ Locus, scales = "free")

```

<div style="margin-bottom:90px;">
</div>

# Complex models

## Choosiness

Here I set preference function as follows:

$1/e^{(x-y)^2/z^2}$

Where 

- $x$, *the most favored courter phenotype*
- $y$, *Courter phenotype*
- $z$, *Promiscuity*, inverse of Choosiness

Draw preference function in several promiscuity

```{r}

for(P in c(3, 4, 5, 6, 9, 15)){
  curve(  
    # Function
    1/exp( ( x^2 )/( P^2 ) ) , 
    # Plot range
    from = -16, to = 16, ylim = c(0,1), 
    main = paste("Promiscuity =", P)
    )
  }

for(P in c(3, 4, 5, 6, 9, 11)){
  curve(  
    # Function
    1-exp( -( x^2 )/( P^2 ) ) , 
    # Plot range
    from = -16, to = 16, ylim = c(0,1), 
    main = paste("Promiscuity =", P)
    )
  }

for(P in c(1:7,9:16)){
  curve(  
    # Function
    1/exp( ( x^2 )/(  (8/(P-8)) ^2 ) ) , 
    # Plot range
    from = -16, to = 16, ylim = c(0,1), 
    main = paste("Pref0 =", P)
    )
  }

```


```{r}

# dat <- load_simul(l = 1, sexsel = "Disruptive", natsel = 0.3)
# 
# ggplot(
#   summarise_simul(dat), 
#   aes(x = Generation, y = Freq, colour = Signal0, group = Signal0)
#   ) +
#   geom_line() +
#   labs(
#     y = "Frequency (%)", colour = "Courter phenotype"
#     )

  # assign(
  #   paste0("plot_d_", locus),
  #   # Density graph
  #   ggplot(dat, aes(x = Signal0, colour = Generation)) +
  #     geom_density() +
  #     labs(
  #       x = "Courter phenotype", 
  #       title = paste(locus, "locus |", cap)
  #       )
  #   )

```

