---
title: "Dynamics"
date: "`r format(Sys.time(), '%Y %m %d')`"
output:
  html_document:
    code_folding: hide
    theme: united
    toc: yes
    toc_float: yes
---

# Setup


Load packages & original functions

Here I set original function `make_genome` that creates genes file


```{r}

rm(list=ls())  # reset workspace

pacman::p_load(tidyverse, knitr, magrittr)     # rmarkdown

opts_chunk$set(prompt=TRUE, message=FALSE, comment="", warning = FALSE) 
options(knitr.kable.NA = '', scipen=100)  # do not show numbers using exponential


#' Title: make_genome
#' Make genes file
#' @param chrom_n: Number of total chromosomes
#' @param g_chrom: Genes/chromosome
#' @param g_total: Total number of genes
#' @param chrom_size: Chromosome size in basepair
make_genome <- function(cn = chrom_n, gt = g_total) {

  set.seed(10) # To make randomization repeatable
  
  gf <- data.frame(
    Chromosome = c(
      1,                      # Sex chromosome
      rep(2:cn, g_chrom) # Autosomal chromosome
      ) %>% 
      sort(),
    # Create random chromosome position to each autosomal gene loci
    position = runif(n = gt, min = 1, max = chrom_size) %>% 
        floor(), # Integers
    # First gene: Sex determination. S vs s. Hemizygous. Freq = 0.5
    Allele1     = c("S", rep("A", gt - 1)),
    Allele1val  = c(1, rep(0.5, gt - 1)),
    Allele2     = c("s", rep("a", gt - 1)),
    Allele2val  = c(0, rep(0, gt - 1)),
    Mode        = c("Hemizygous", rep("Additive", gt - 1)),
    Freqal1pop1 = c(0.5, rep(1.0, gt - 1)),
    Freqal1pop2 = c(0.5, rep(0, gt - 1))
    ) %>% 
    mutate(name = str_c("chr", Chromosome, "_", position, sep = "")) %>% 
    dplyr::select(name, Chromosome:Freqal1pop2)
  
  return(gf)
  
  }


#' Title: make_config
#' Make simulation configuration file
#' @param gset: Genome setting. It has to be either [dis_unlink, dis_link, quant_learn, quant]
#' @param sexsel: Sexual selection style
#' @param gen: Generations
#' @param mig: Migration in each >1 generation
#' @param maten: Sampled mate
make_config <- function(
  gset = "", sexsel = "", bdmi = "", h = "", gen = 100, mig = 0, maten = 50
  ) {

  head <- paste(
    "RandomSeed\t=\t.1356",         
    paste0("MarkerFile\t=\t/lustre/k.atsumi/Config/markers_", gset, ".txt"),
    # "MarkerFile\t=\tmarkers.txt",    #Marker File Containing marker info
    paste0("MarkerProbFile\t=\t/lustre/k.atsumi/Config/recomb_", gset, ".txt"),
    # "MarkerProbFile\t=\trecomb.txt", #Recombination probabilities
    "UseUniformRec\t=\tyes",		# ignore predefined breakpoints above, use uniform.
    "IgnoreMarkerFreq\t=\tyes", #setting this option to yes will treat all markers as ancestry informative, regardless of their frequency specified in the markers file
    # "DumpNatSelProb\t=\tOff", 
    paste0("OutputFolder\t=\t/lustre/k.atsumi/Result/", gset, "_", sexsel, "_b", bdmi, "h", h, "/0"), #Output folder
    sep = "\n"
    )
  
  mid <- if(gset == "quant") {
    #Dynamic genes & phenotype files
    paste(
      "GeneFile\t=\t/lustre/k.atsumi/Config/genes_quant_random0.txt",
      "PhenotypeFile\t=\t/lustre/k.atsumi/Config/phenotypes_quant_random0.txt",
      paste0("NaturalSelection\t=\t/lustre/k.atsumi/Config/naturalsel_quant_b", bdmi, "h", h, ".txt"),                     #Natural selection file
      sep = "\n"
      )
  } else {
    #Static genes & phenotype files
    paste(
      paste0("GeneFile\t=\t/lustre/k.atsumi/Config/genes_", gset, ".txt"),
      paste0("PhenotypeFile\t=\t/lustre/k.atsumi/Config/phenotypes_", gset, ".txt"),
      paste0("NaturalSelection\t=\t/lustre/k.atsumi/Config/naturalsel_b", bdmi, "h", h, ".txt"),                     #Natural selection file
      sep = "\n"
      )
  }
  
  last <-  paste(
    paste0("SexualSelection\t=\t/lustre/k.atsumi/Config/sexualsel_", sexsel, ".txt"), #Sexual selection file
    #population names:
    "pop2_name\t=\tp2",
    "pop2_ancestry_label\t=\tp2",
    "hybrid_name\t=\thyb",
    "pop1_name\t=\tp1",
    "pop1_ancestry_label\t=\tp1",
    #Pop sizes
    "pop1_init_size\t=\t500",
    "pop1_size_limit\t=\t500",
    "pop2_init_size\t=\t500",
    "pop2_size_limit\t=\t500",
    "hybrid_size_limit\t=\t2000",
    #Sex ratio
    "pop1_male_ratio\t=\t0.5",
    "pop2_male_ratio\t=\t0.5",
    #Migration
    "migration_only_first_gen\t=\tyes",
    #migration first generation:
    "gen0_pop2_to_hybrid\t=\t500",
    "gen0_pop1_to_hybrid\t=\t500",
    #migration each gen:
    paste0("pop2_to_pop1\t=\t", mig),
    paste0("pop2_to_pop1\t=\t", mig),
    paste0("pop2_to_hybrid\t=\t", mig),
    paste0("hybrid_to_pop2\t=\t", mig),
    paste0("hybrid_to_pop1\t=\t", mig),
    paste0("pop1_to_pop2\t=\t", mig),
    paste0("pop1_to_hybrid\t=\t", mig),
    paste0("generations\t=\t", gen),
    "avg_female_gamete\t=\t10",
    "std_female_gamete\t=\t3.16",
    "kids_per_female_func\t=\tPoisson",
    "samplefreq\t=\t1", 
    # "samplegens\t=\t/lustre/k.atsumi/Config/samplegens.txt", 
    "MarkerOutput\t=\tOff", #On or Off
    paste0("SampleMate\t=\t", maten),
    "NumThreads\t=\t1",
    "DisableOutputForFirst3Gens\t=\tyes",
    sep = "\n"
    )
  
  return( paste(head, mid, last, sep = "\n") )
  
  }


#' Title: make_bash
#' Make simulation bash file
#' @param gset: Genome setting. It has to be either [dis_unlink, dis_link, quant_learn, quant]
#' @param sexsel: Sexual selection style
#' @param gen: Generations
#' @param mig: Migration in each >1 generation
#' @param maten: Sampled mate
make_bash <- function(gset = "", sexsel = "", s = "") {

  f = paste0("./Bash/", gset, "_", sexsel, ".sbatch.sh") #Output file name
    
  paste(
    "#!/bin/bash",
    paste("#PBS -N hyb", sexsel, gset, sep = "_"),
    "#PBS -q blade",
    "#PBS -l nodes=1:ppn=16,mem=128gb,walltime=48:00:00",
    "module load gcc",
    "# Repetition 1 to 100",
    "REPS=101",
    "# X: first random seed value, its decimal because random seeds are between 0 and 1. If there is an existing one, let X be the newest one.",
    "# Y: new randmon seed value; let it have the same value as X here",    
    "X=.1356",
    "Y=.1356",
    "# COUNTERA, first iteration name; COUNTERB, new iteration name",
    "COUNTERA=0",
    "COUNTERB=0",
    "while [ $COUNTERA -lt $REPS ]; do",
    " echo \"$COUNTERA\"",
    " Y=`echo \"$Y + 0.0001\" | bc`",
    " let COUNTERB=COUNTERB+1",
    " # Change random seed ($X->$Y) and output folder (${COUNTERA}->${COUNTERB})",
    " # Run simulation",
    sep = "\n"
    ) %>% 
    cat(., "\n", append = FALSE, file = f) #Renew file
  
  #Repeat for each combinaiton of incompatibilities (BDMI, Underdominance)
  for(sb in s){   #BDMI
    for(sh in s){ #Underdominance - against Heterozygotes
      
      if(gset == "quant") {  #Change Genes & Phenotype files used
        paste0(" sed -i -e \"s/quant_random$X/quant_random$Y/g\" /lustre/k.atsumi/Config/quant_", sexsel, "_b", sb, "h", sh, ".cfg") %>%
          cat(., "\n", append = TRUE, file = f)
      }
      
      paste(
        # Change random seed ($X->$Y) and output folder (${COUNTERA}->${COUNTERB})
        paste0(" sed -i -e \"s/$X/$Y/g\" -e \"s/rep${COUNTERA}dir/rep${COUNTERB}dir/g\" /lustre/k.atsumi/Config/", gset, "_", sexsel, "_b", sb, "h", sh, ".cfg"),
        # Run simulation
        paste0(" /lustre/k.atsumi/admixem/bin/admixemp /lustre/k.atsumi/Config/", gset, "_", sexsel, "_b", sb, "h", sh, ".cfg &"),
        sep = "\n"
        ) %>%
        cat(., "\n", append = TRUE, file = f)
    }}

  paste(
    " wait",
    " Make new random seed and iteration value",
    " X=`echo \"$X + 0.0001\" | bc`",
    " let COUNTERA=COUNTERA+1",
    "done",
    sep = "\n"
    ) %>%
    cat(., "\n", append = TRUE, file = f)
  
  }
  
```



# Simple model settings

- With sex chromosome 1
  - Allele (S) with trait value 1 will be male

- BDMI between Chrom 2 & 3

- Underdominance at chrom 4

- Migration happens in only first generation

- Max 2000 indv for hybrid pop

- Mate sampling: 50 indv

## Natural selection

```{r}

# Variation in selection strength
nats <- c(0, 0.1, 0.4, 0.8)

for(bdmi in nats){
  for(ud in nats){
    paste(
      "Population\tGen\tSelection",
      "species1\t-1\t1",
      "species2\t-1\t1",
      paste0(
        "hyb\t-1\tb=", 
        bdmi, 
        ", u=", 
        ud, 
        ", if(BDMI==30 && Underdominance==0.5, 1-b-u+b*u, if(BDMI==30 && Underdominance!=0.5, 1-b, if(BDMI!=30 && Underdominance=0.5, 1-u, 1)))"  
        ),
      sep = "\n"
      ) %>% 
      write(paste0("./Config/naturalsel_b", bdmi, "h", ud, ".txt"))
}}

```



## Gene file

```{r}

chrom_size <- 1*10^5
paste("Chromosone size (bp):", chrom_size)

g_chrom <- 2
paste("Genes per autosome:", g_chrom)

g_trait_n <- 1    # N of trait loci
paste(g_trait_n, "locus determine courter trait")

g_pref_n  <- 1     # N of preference loci
paste(g_pref_n, "locus determine mate preference peak, with constant choosiness")
# trait_n   <- 1      # N of courter trait

chrom_n <- 1 + 2 + 1 + 10 + g_trait_n + g_pref_n #Sex, BDMI, Underdominance, Hybrid index, Trait, Preference (Maximum number)
paste("Chromosone number (with sex chrom):", chrom_n)

g_total <- 1 + (chrom_n - 1)*g_chrom
paste("Total number of genes:", g_total)

# Create gene file
gf <- make_genome()
# Epistasis
gf[2, "Allele1val"] <- 5
gf[2, "Allele2val"] <- 0
gf[2+g_chrom, "Allele1val"] <- 0
gf[2+g_chrom, "Allele2val"] <- 10

write_tsv(gf, "./Config/genes_dis_link.txt")
file.copy(
  "./Config/genes_dis_link.txt", "./Config/genes_dis_unlink.txt", overwrite = T
  )

```


## Phenotype files

Create `phenotypes_{genome setting}.txt`. I made 3 phenotype files with genome setting
- *dis_unlink* : Trait determined by 1 locus, which is unlinked with preference genes
- *dis_link* : Trait determined by 1 locus, which is same gene determining peak

```{r}

Pheno <- data.frame(
  Phenotypes = c("Sex", "Pref0", "Pref1", "Signal0", "Hybridindex", "BDMI", "Underdominance"), 
  Formula    = c(
    gf[1, "name"],   # Sex
    gf[2+g_chrom*13, "name"], # Preference (peak)
    gf[2+g_chrom*13, "name"], # Preference
    "Signal",      # Signal, determined later according to link/unlinked with preference
    paste0(gf$name[(2+g_chrom*3):(3+g_chrom*12)], collapse = "+"),  # Hybrid Index
    paste0(gf[2, "name"], "+", gf[2+g_chrom, "name"]), # BDMI
    gf[2+g_chrom*2, "name"]  # Underdominance
    )
  )

# dis_unlink
Pheno[6,2] <- gf[2+g_chrom*14, "name"]
write_tsv(Pheno, "./Config/phenotypes_dis_unlink.txt")

# dis_link
Pheno[6,2] <- gf[2+g_chrom*13, "name"]
write_tsv(Pheno, "./Config/phenotypes_dis_link.txt")


for(genomeset in c("dis_unlink", "dis_link")){
  for(ss in c("speriorhyb", "disruptive", "flip", "heterorandom", "pdominance", "recessive", "selfref")){
    make_bash(gset = genomeset, sexsel = ss, s = nats)
    
    for(BDMI in nats){
      for(H in nats){
        make_config(gset = genomeset, sexsel = ss, bdmi = BDMI, h = H) %>% 
          write(paste0("./Config/", genomeset, "_", ss, "_b", BDMI, "h", H, ".cfg"))
  }}}}

```

By running file `config_marker_~~.txt` on Admix'em, marker file (`markers.txt`) and recombination file (`recomb.txt`) will be made

```{r}

m_chrom <- 2 #Marker per chromosome

paste(
  "1",        #Generate markers
  "0.99",     #Random seed
  chrom_n,    #N of Chromosome 
  sep = "\n"  #改行
  ) %>% 
  cat(
    ., "\n", file = "./Config/config_marker_dis_link.txt", 
    append = FALSE  #Renew file
    )

for(c in 1:chrom_n){　#Repeat chromosome size
  cat(
    chrom_size, "\n", file = "./Config/config_marker_dis_link.txt",
    append = TRUE   #Add to existence file
    )
  }

paste(
  chrom_n*m_chrom,   #N of markers
  "1.0",               #Pop 1 allele freq
  0,                 #Pop 1 allele freq SD
  "1.0",               #Pop 2 allele freq
  0,                 #Pop 2 allele freq SD
  "markers_dis_link.txt",     #marker file name
  "y",                 #Reconbination file
  4,                 #Reconbination N
  5,                 #Male female difference
  "recomb_dis_link.txt",      #recombination file
  sep = "\n"
  ) %>% 
  cat(., "\n", file = "./Config/config_marker_dis_link.txt", append = TRUE)

file.copy("./Config/recomb_dis_link.txt","./Config/recomb_dis_unlink.txt", overwrite = T)
file.copy("./Config/markers_dis_link.txt","./Config/markers_dis_unlink.txt", overwrite = T)

```






# Quantitative trait & Learned preference models

Preference is determined not by gene, but by phenotypic distribution of the hybrid population

Choosiness, Peak preference, Traits & Hybrid index are determined by 50 loci, respectively

```{r}

chrom_size <- 1*10^7
paste("Chromosone size (bp):", chrom_size)

chrom_n <- 1 + 25
paste("Chromosone number (with sex chrom):", chrom_n)

g_chrom <- 18
paste("Genes per autosome:", g_chrom)

g_total <- 1 + (chrom_n - 1)*g_chrom
paste("Total number of genes:", g_total)

# Create gene file
gf_org <- make_genome()
gf_org$Allele1val <- c(1, rep(0.5, nrow(gf_org)-1))
gf_org$Allele2val <-0

Pheno_org <- data.frame(
  Phenotypes = c("Sex", "Pref0", "Pref1", "Signal0", "Underdominance", "BDMI", "Hybridindex"), 
  Formula    = c(
    gf_org[1, "name"],   # Sex
    gf_org[2+g_chrom*13, "name"], # Preference (peak)
    gf_org[2+g_chrom*13, "name"], # Preference
    rep(NA,4)
    )
  )

# Make phenotype files with randomized positions of functional genes
for(rseed in 0:100) {
  
  Pheno <- Pheno_org #Make copy of original object
  gf <- gf_org       #Make copy of original object

  set.seed(rseed+1)
  #Randomly set incompatibility loci & Neutral loci (Hybrid index)
  gf$Allele1val[sample(2:nrow(gf),200)] <- c(
    rep(10, 50), #Hybrid index
    rep(3, 50), #Underdominance loci -> Fitness reduction if trait value is 3
    rep(5, 50), #BDMI loci -> Fitness reduction if trait value is 24 
    rep(0, 50)  #BDMI loci 
    )
  gf$Allele2val[gf$Allele1val == 0] <- 7 #BDMI loci -> Fitness reduces if trait value is 24 

  write_tsv(gf, paste0("./Config/genes_quant_random", rseed, ".txt"))
  
  #Hybrid index
  Pheno[7,2] <- paste0( 
    "(", paste0(gf$name[gf$Allele1val == 10], collapse = "+"), ")/500"
    )
  #Underdominance
  Pheno[5,2] <- paste0("if(", gf$name[gf$Allele1val == 3], "==3,1,0)", collapse = "+")
  #BDMI
  Pheno[6,2] <- paste0("if(", gf$name[gf$Allele1val == 5], "+", 
           sample(gf$name[gf$Allele2val == 7],50), "==24,1,0)", #Shuffle 
           collapse = "+")
  
  gf_ni <- filter(gf, Allele1val == 0.5, Allele2val == 0) #Non-incompat & Non-HI loci
  
  #Trait & Preference loci
  set.seed(rseed+101) # To make randomization repeatable
  Pheno[2,2] <- paste0(sample(gf_ni$name, 50), collapse = "+")
  set.seed(rseed+201) # Random genes, while differentiating with above
  Pheno[3,2] <- paste0(sample(gf_ni$name, 50), collapse = "+")
  set.seed(rseed+301) # Random genes, while differentiating with above
  Pheno[4,2] <- paste0(sample(gf_ni$name, 50), collapse = "+")
  
  write_tsv(Pheno, paste0("./Config/phenotypes_quant_random", rseed, ".txt"))
  
}

```

By running this file on Admix'em, marker file (`markers.txt`) and recombination file (`recomb.txt`) will be made

```{r}

m_chrom <- 2 #Marker per chromosome

paste(
  "1",        #Generate markers
  "0.99",     #Random seed
  chrom_n,    #N of Chromosome 
  sep = "\n"  #改行
  ) %>% 
  cat(
    ., "\n", file = "./Config/config_marker_quant.txt", append = FALSE  #Renew file
    )

for(c in 1:chrom_n){　#Repeat chromosome size
  cat(
    chrom_size, "\n", file = "./Config/config_marker_quant.txt",
    append = TRUE   #Add to existence file
    )
  }

paste(
  chrom_n*m_chrom,   #N of markers
  "1.0",               #Pop 1 allele freq
  0,                 #Pop 1 allele freq SD
  "1.0",               #Pop 2 allele freq
  0,                 #Pop 2 allele freq SD
  "markers_quant.txt",     #marker file name
  "y",                 #Reconbination file
  8,                 #Reconbination N
  5,                 #Male female difference
  "recomb_quant.txt",      #recombination file
  sep = "\n"
  ) %>% 
  cat(., "\n", file = "./Config/config_marker_quant.txt", append = TRUE)

```

## Natural selection

```{r}

nats <- c(0.05, 0.01, 0.005, 0) # Variation in selection strengthw

for(bdmi in nats){
  for(ud in nats){
    paste(
      "Population\tGen\tSelection",
      "species1\t-1\t1",
      "species2\t-1\t1",
      paste0(
        "hyb\t-1\tb=", 
        bdmi, 
        ", u=", 
        ud, 
        ", pow(1-b,BDMI) * pow(1-u,Underdominance)"  
        ),
      sep = "\n"
      ) %>% 
      write(paste0("./Config/naturalsel_quant_b", bdmi, "h", ud, ".txt"))
}}

```


```{r}

# Config file
for(ss in c("obl_posit", "obl_negat", "constchoosy", "hybridrandom")){
  make_bash(gset = "quant", sexsel = ss, s = nats) %>% 
      write(paste0("./Bash/", "quant_", ss, ".sbatch.sh"))

  for(BDMI in nats){
    for(H in nats){
      make_config(gset = "quant", sexsel = ss, bdmi = BDMI, h = H) %>% 
        write(paste0("./Config/quant_", ss, "_b", BDMI, "h", H, ".cfg"))
    }}}

```

- *quant_unlink* : Trait determined by independent 16 loci which are unlinked with preference genes

<div style="margin-bottom:90px;">
</div>

# Complex models

## Choosiness

Here I set preference function as follows:

$1/e^{(x-y)^2/z^2}$

Where 

- $x$, *the most favored courter phenotype*
- $y$, *Courter phenotype*
- $z$, *Promiscuity*, inverse of Choosiness

Draw preference function in several promiscuity

### Positive learning model & Constant choosiness model

```{r}

for(P in c(5, 7, 9, 11, 15)){
  curve(  
    # Function
    1/exp( ( x^2 )/( P^2 ) ) , 
    # Plot range
    from = -50, to = 50, ylim = c(0,1), 
    main = paste("Promiscuity =", P)
    )
}

```

### Negative learning model

```{r}

for(P in c(17, 19, 21, 23)){
  curve(  
    # Function
    1-exp( -( x^2 )/( P^2 ) ) , 
    # Plot range
    from = -50, to = 50, ylim = c(0,1), 
    main = paste("Promiscuity =", P)
    )
  }

```


### Choosiness reduction model

```{r}

# Choosiness reduction model
for(P in c(1, 5, 10, 15, 20, 24)){   #Courter genotype
  a <- 90
  curve(  
    # Function
    1/exp( ( x^2 )/(  (a/(P-25)) ^2 ) ) , 
    # Plot range
    from = -50, to = 50, ylim = c(0,1), 
    main = paste("Pref0 =", P)
    )
  }

```



<div style="margin-bottom:70px;">
</div>

# After simulation, before analysis

Remove unused result files

```
cd /lustre/k.atsumi
# Remove all marker, genes, natselprobdump files in the subdirectories
find ./ -name "*markers.txt" -delete        #Marker files
find ./ -name "*natselprobdump.txt" -delete #? Unused fle
find ./ -name "*genes.txt" -delete          #Genes files
```


# Old codes

```{r, include=FALSE}


```
